src/analysis/ml.py

```diff
@@ -152,27 +175,92 @@ def plot_dendrogram(model, **kwargs):
     # Plot the corresponding dendrogram
     dendrogram(linkage_matrix, **kwargs)
 
-
-def hac(dataframe: pd.DataFrame, output_png: str, output_file: str):
+def hac_vertical(dataframe: pd.DataFrame, output_png: str, output_file: str, linkage='ward', metric='euclidean'):
     """
     Performs hierarchical agglomerative clustering on the dataframe,
-    creates a dendrogram of the resulting tree,
+    creates a dendrogram of the resulting tree USING SEABORN and SCKIT-LEARN for the cluster groups,
     and saves the dendrogram and the cluster labels of said dendrogram in separate files.
     @param dataframe: binary dataframe of edge comparison between algorithms from summarize_networks
     @param output_png: the file name to save the dendrogram image
     @param output_file: the file name to save the clustering labels
     """
+
+    if linkage not in linkage_methods:
+        raise ValueError(f"linkage={linkage} must be one of {linkage_methods}")
+    if linkage == "ward":
+        if metric != "euclidean":
+            print("For linkage='ward', the metric must be 'euclidean'; setting metric = 'euclidean")
+            metric = "euclidean"
+    if metric not in distance_metrics:
+        raise ValueError(f"metric={metric} must be one of {distance_metrics}")
+        
     X = dataframe.reset_index(drop=True)
-    X = X.transpose()
-    model = AgglomerativeClustering(distance_threshold=0.5, n_clusters=None)
+    column_names = X.head()
+    column_names = [element.split('-')[1] for element in column_names]
+    X = X.transpose() 
+
+    # creating the colors per algorithms
+    custom_palette = sns.color_palette("tab10", len(column_names))
+    label_color_map = {label: color for label, color in zip(column_names, custom_palette)}
+    row_colors = pd.Series(column_names, index=X.index).map(label_color_map)
+
+    #plotting the seaborn figure
+    plt.figure(figsize=(10, 7))
+    clustergrid = sns.clustermap(X, metric=metric, method=linkage, row_colors=row_colors, col_cluster=False)
+    clustergrid.ax_heatmap.remove()
+    clustergrid.cax.remove()
+    clustergrid.ax_row_dendrogram.set_visible(True)
+    clustergrid.ax_col_dendrogram.set_visible(False)
+    legend_labels = [plt.Rectangle((0, 0), 0, 0, color=label_color_map[label]) for label in label_color_map]
+    plt.legend(legend_labels, label_color_map.keys(), bbox_to_anchor=(1.02, 1), loc='upper left')
+    
+    make_required_dirs(output_png)
+    plt.savefig(output_png, bbox_inches="tight")
+
+    # getting the label of which group each algorithm combination falls under
+    model = AgglomerativeClustering(linkage=linkage, affinity=metric,distance_threshold=0.5, n_clusters=None)
     model = model.fit(X)
```

Comment:
We have to run clustering twice? We can't get this information from the seaborn/scipy clustering? That seems inefficient for large collections of graphs.

Reply:
Here is some relevant discussion:
- https://stackoverflow.com/questions/56526707/extract-rows-of-clusters-in-hierarchical-clustering-using-seaborn-clustermap
- https://stackoverflow.com/questions/27924813/extracting-clusters-from-seaborn-clustermap

Reply:
For clustermap, even though it generate the hierarchical clustering and displays the dendrogram, it does not directly return the cluster assignments of the data. 

Rather than using the AgglomerativeClustering to calculate the cluster groups, we can use fcluster from the scipy library. 
1. we would grab the row cluster grid object, which gives a linkage matrix from the sns.clustermap dendrogram
2. we then have to use fcluster
- but it takes in an argument t that cuts the dendrogram at a specific distance to form clusters

` # Use linkage matrix from seaborn clustergrid to generate cluster assignments
    linkage_matrix = clustergrid.dendrogram_row.linkage
    # t = the maximum inter-cluster distance allowed
    clusters = fcluster(linkage_matrix, t=0.5, criterion='distance')`

Or we can get rid of it all together

Reply:
Do we know whether the fcluster t argument has the same role as the sklearn AgglomerativeClustering distance_threshold argument? distance_threshold is currently set to 0.5 in our code. If they have same behavior, then using fcluster and setting t=0.5 (or making it another option in the config file) might make sense.

Reply:
Yeah, the fcluster takes an argument t which is the distance threshold, which is the same (theoretically) as the distance_threshold in AgglomerativeClustering. I made both 0.5, and the way the groupings were identical so that made sense in my mind. The only difference right now is that fcluster is starting the numbering of the groups at 1, and AgglomerativeClustering starts the numberings of groups at 0 (but I am going to see if I can make fcluster start at 0 as well)

Reply:
the numbering differs between vertical and horizontal now, however the groupings are correct. So, using fcluster does work